File: BC-W4/Static/Identifiable.cpp
#include <iostream>

class Identifiable {
    private:
        int id;
    public:
        static int lastId;
    public:
        Identifiable() {
            lastId += 1;
            this->id = lastId;
        }
        
        Identifiable(const Identifiable& copy) {
            lastId += 1;
            this->id = lastId;
        }
        
        Identifiable& operator=(const Identifiable& copy) {
            return *this;
        }
        
        ~Identifiable() {}
        
        int getLastId() {
            return this->id;
        }
};

int Identifiable::lastId = 0;

int main() {
    Identifiable *count1 = new Identifiable();
    Identifiable *count2 = new Identifiable();
    Identifiable *count4 = new Identifiable(*count2);
    Identifiable *count3 = new Identifiable();
    *count3 = *count1;
    
    std::cout <<"1:" << count1->getLastId() << std::endl;
    std::cout <<"2:"<< count2->getLastId() << std::endl;
    std::cout <<"3:"<< count3->getLastId() << std::endl;
    std::cout <<"4:"<< count4->getLastId() << std::endl;
    
    delete count1;
    
    // std::cout <<"1:" << count1->getLastId() << std::endl;
    std::cout <<"2:"<< count2->getLastId() << std::endl;
    std::cout <<"3:"<< count3->getLastId() << std::endl;
    std::cout <<"4:"<< count4->getLastId() << std::endl;
    return 0;
}

File: BC-W4/PHP/Identifiable/Identifiable.php
<?php
    class Identifiable {
        private $id;
        private static $lastId  = 0;
        
        public function __construct() {
            self::$lastId += 1;
            $this->id = self::$lastId;
        }
        
        public function __get($name) {
            if ( property_exists($this, $name) ) {
                return $this->$name;
            }
            throw new Exception("Attribute error: attribute $name not found");
        }
    }
    
    $count1 = new Identifiable();
    $count2 = new Identifiable();
    $count3 = new Identifiable($count2);
    $count4 = new Identifiable();
    $count4 = $count1;
    
    echo "1:" . $count1->id . PHP_EOL;
    echo "2:" . $count2->id . PHP_EOL;
    echo "3:" . $count3->id . PHP_EOL;
    echo "4:" . $count4->id . PHP_EOL;
    
?>